libraries/AP_HAL_Linux/GPIO_RPI.cpp

```diff
@@ -19,85 +20,186 @@
+    _gpio[pin / pins_per_register] |= mask;
+}
+
+void GPIO_RPI::set_gpio_mode_in(int pin)
+{
+    // Each register can contain 10 pins
+    const uint8_t pins_per_register = 10;
+    // Calculates the position of the 3 bit mask in the 32 bits register
+    const uint8_t tree_bits_position_in_register = (pin%pins_per_register)*3;
+    // Create a mask that only removes the bits in this specific GPIO pin, E.g:
+    // 0b11'111'111'111'111'111'111'000'111'111'111 for the 4th pin
+    const uint32_t mask = ~(0b111<<tree_bits_position_in_register);
+    // Apply mask
+    _gpio[pin / pins_per_register] &= mask;
+}
+
+void GPIO_RPI::set_gpio_mode_out(int pin)
+{
+    // Each register can contain 10 pins
+    const uint8_t pins_per_register = 10;
+    // Calculates the position of the 3 bit mask in the 32 bits register
+    const uint8_t tree_bits_position_in_register = (pin%pins_per_register)*3;
+    // Create a mask that enable the bit the sets output functionality
+    // 0b00'000'000'000'000'000'000'001'000'000'000 enables output for the 4th pin
+    const uint32_t mask = 0b001 << tree_bits_position_in_register;
+    // Apply mask
+    _gpio[pin / pins_per_register] |= mask;
+}
+
+void GPIO_RPI::set_gpio_high(int pin)
+{
+    // Calculate index of the array for the register GPSET0 (0x7E20'001C)
+    constexpr uint32_t gpset0_memory_offset_value = 0x1c;
+    constexpr uint32_t gpset0_index_value = gpset0_memory_offset_value / sizeof(*_gpio);
+    _gpio[gpset0_index_value] = 1 << pin;
+}
+
+void GPIO_RPI::set_gpio_low(int pin)
+{
+    // Calculate index of the array for the register GPCLR0 (0x7E20'0028)
+    constexpr uint32_t gpclr0_memory_offset_value = 0x28;
+    constexpr uint32_t gpclr0_index_value = gpclr0_memory_offset_value / sizeof(*_gpio);
+    _gpio[gpclr0_index_value] = 1 << pin;
+}
+
+bool GPIO_RPI::get_gpio_logic_state(int pin)
+{
+    // Calculate index of the array for the register GPLEV0 (0x7E20'0034)
+    constexpr uint32_t gplev0_memory_offset_value = 0x34;
+    constexpr uint32_t gplev0_index_value = gplev0_memory_offset_value / sizeof(*_gpio);
+    return _gpio[gplev0_index_value] & (1 << pin);
+}
+
+uint32_t GPIO_RPI::get_address(GPIO_RPI::Address address, GPIO_RPI::PeripheralOffset offset) const
+{
+    return static_cast<uint32_t>(address) + static_cast<uint32_t>(offset);
+}
+
+volatile uint32_t* GPIO_RPI::get_memory_pointer(uint32_t address, uint32_t range) const
+{
+    auto pointer = mmap(
+        nullptr,                         // Any adddress in our space will do
+        range,                           // Map length // TODO: check this value
+        PROT_READ|PROT_WRITE|PROT_EXEC,  // Enable reading & writting to mapped memory
+        MAP_SHARED|MAP_LOCKED,           // Shared with other processes
+        _system_memory_device,           // File to map
+        address                          // Offset to GPIO peripheral
+    );
+
+    if (pointer == MAP_FAILED) {
+        AP_HAL::panic("Can't open /dev/mem");
```

Comment:
This panic should be in the caller, if this function returns nullptr. Also, maybe adjust the message string to be more clear about mmap failing.